### PR TITLE
Assume login feature

### PR DESCRIPTION
**Pull request by KevinVdV**
## References
_Add references/links to any related tickets or PRs. These may include:_
* Link to [REST Contract](https://github.com/DSpace/Rest7Contract/pull/109)

## Description
Rest api implementation for the assume login feature. Included are the actual "login as" implementation by making use of the "X-On-Behalf-Of" header & code to support "loginOnBehalfOf" on the "features" feature.

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* StatelessAuthenticationFilter: contains the code for the actual "login as"
* LoginOnBehalfOfFeature: contains the code for the feature

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java:63]: this cannot be an instance attribute otherwise there are issue when multiple login request are processed at the same time

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java:37]: it is not really needed, we can simplify the code here

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java:112]: the turnOff is not really needed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java:122]: I would suggest to don't revert the configuration manually here the change as this is already done in a Before method (just to avoid in future confusion in other developers)

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/LoginAsEPersonIT.java:149]: I would like to have an additional test using the loginAs feature in some core functionalities like
- workspaceitems (to check that the item created with the loginAs are owned by the impersonated user or that the user can search for his items)
- tasks are claimed according to the loginas
- actions reserved to an admin are prevent to the loginAs (for instance trying to delete an archived item after loginas a normal user)

**Review by abollini — CHANGES_REQUESTED** [review]
I have added just minor suggestions about the IT. The only coding issue is related to the use of an instance attribute in the Spring Security Filter that should be removed (maybe refactoring the code to use a custom exception)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java:88]: This `isErrorOnBehalfOf` flag is really odd to me & doesn't seem to be necessary.  It seems like you should be able to refactor this code to no longer use it....ideally you should return immediately after any call to `response.sendError()` (to avoid anything else trying to add to the response).  So, maybe this method could be refactored to simply check for the header here, and call `getOnBehalfOfAuthentication`.  If that method throws an error, catch it and call `sendError()`.  Otherwise, proceed as normal.

**Review by tdonohue — COMMENTED** [review]
@KevinVdV : Besides what @abollini has already noted (which I agree with), I added one other request about simply refactoring the filter to remove the `isErrorOnBehalfOf` flag altogether (as I agree it's fragile, and it's not really necessary anyways).

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java
Comment (KevinVdV) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java]: @abollini We can't test on an empty list here as some features are always returned by default. For example the "AlwaysTrueFeature". As we can't anticipate which features will be returned it would be best to check for an exclusion.

**Comment by KevinVdV** [comment]
@tdonohue @abollini We have processed your test feedback, let us know if there is anything else or if we missed something.

**Review by benbosman — APPROVED** [review]
I've deployed and tested this branch, and I couldn't see any problems.
I was able to verify:
* Without `webui.user.assumelogin = true` the `X-On-Behalf-Of` header is denied
* With `webui.user.assumelogin = true` the pages return the content based on the user's permissions, including e.g. the workflow tasks, or claiming a workflow task
* With `webui.user.assumelogin = true` the pages return forbidden for pages the user is not authorized to use

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/LoginOnBehalfOfFeature.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/LoginOnBehalfOfFeature.java:67]: Could we add some comments to this method. It's not well described currently.  It looks like it's covering some scenarios that are not described in the class JavaDocs.  Namely, it looks like you have LoginOnBehalfOf  rights if:
1. Feature is enabled, you request it on a Site and you are an admin
2. Or, Feature is enabled, you request it on an EPerson *and* you are NOT that EPerson and that EPerson is *NOT* and Admin.

Am I reading the code correctly?  The logic here is a bit hard to read, so it'd be good to describe in a comment. or two.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java:87]: This entire `Pair` logic seems really odd to me.  Why not just have `getAuthentication()` return either `Authentication` or null?  If an error needs to occur, then `getAuthentication()` should throw an exception, which then gets caught in this method and turned into a `response.sendError()`.

I don't see a scenario where you'd ever want to return *both* an `Authentication` and an error (which is what this `Pair` logic seems to imply could occur).  So, there's no need to have this `Pair`, as either you have an Authentication, it's null, or you throw an error.  Those can all be handled just via normal error handling scenarios.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java:181]: This also seems a bit odd to me.  Why would you not return an error when an error occurred?  This seems to be essentially a no-op if a SQLException happens (or am I missing something)?

**Review by tdonohue — COMMENTED** [review]
@KevinVdV : I gave this another review today.  While I think it's looking better, the `Pair` logic seems really odd to me. I'd like to either better understand why you decided on this route (which means better code commenting here), or replace it with normal error handling.  See more inline below.

**Comment by KevinVdV** [comment]
@tdonohue We have processed your feedback, can you review again ?

**Review by abollini — COMMENTED** [review]
thanks @Raf-atmire to have addressed my feedback. I have just added a small cleanup comment inline

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java
Comment (KevinVdV) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java]: Good catch, we removed it.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/StatelessAuthenticationFilter.java:117]: These JavaDocs are now incorrect as they describe how the method worked prior to the refactor. The `Pair` was removed.  @KevinVdV could we get these updated quickly?  This is the last thing I see that needs updating before we can merge.

**Review by abollini — APPROVED** [review]

**Comment by abollini** [comment]
all my feedback have been processed. Can be merged once @tdonohue is satisfied

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now. Thanks @KevinVdV and @Raf-atmire

**Pull request merged**
